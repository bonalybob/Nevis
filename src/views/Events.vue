<template>
  <main>
    <h1 class="mx-10 mb-1">
      <back-arrow to="/" />Events
    </h1>
    <div class="mx-12 mb-3">
      <router-link tag="button" class="button" to="/events/create">Create Event</router-link>
      <button v-if="$archive.connected" class="button" @click="importArchive">Import Archive</button>
      <button class="button" @click="getEvents">Refresh</button>
    </div>
    <div v-if="events && events.length > 0" class="shadow mx-12 mb-3 p-2">
      <table class="w-full font-body">
        <tr class="font-heading text-center hover:bg-blue-light">
          <th>Event Name</th>
          <th>Event Date</th>
        </tr>
        <router-link
          v-for="event of events"
          :key="event.id"
          :to="'/events/' + event.id"
          class="text-center even:bg-blue-lightest hover:bg-blue-light"
          tag="tr"
        >
          <td>{{ event.name }}</td>
          <td>{{ event.date }}</td>
        </router-link>
      </table>
    </div>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'

export default {
  components: {
    'back-arrow': BackArrow,
  },

  data: function () {
    return {
      events: [],
    }
  },

  mounted: function () {
    if (this.$database.connection === null || !this.$database.connected) {
      this.$router.push('/')
      this.$messages.addMessage('Problem Connecting To Database', 'error')
    }
    else {
      this.createTables()
      this.getEvents()
      this.createArchiveConnection()
    }
  },

  methods: {
    getEvents: function () {
      return this.$database.query('SELECT * FROM events ORDER BY date DESC')
        .then(result => { this.events = result })
        .catch(() => this.$messages.addMessage('Problem Fetching Events', 'error'))
    },

    createArchiveConnection: function () {
      this.$archive.connection = this.$mysql.createConnection({
        host: this.$database.connection.config.host,
        port: this.$database.connection.config.port,
        user: this.$database.connection.config.user,
        password: this.$database.connection.config.password,
        database: 'archive',
      })
      return this.$archive.connect()
        .then(() => { this.$archive.connected = true })
        .then(() => {
          return this.$archive.query(`CREATE TABLE IF NOT EXISTS people(
          id INT AUTO_INCREMENT PRIMARY KEY,
          status TEXT,
          name TEXT,
          siid TEXT,
          membershipNumber TEXT,
          gender TEXT,
          club TEXT,
          yearOfBirth INT)`)
        })
        .catch(() => this.$messages.addMessage('Problem Connecting To The Archive', 'error'))
    },

    importArchive: async function () {
      const { dialog } = this.$electron.remote
      return dialog.showOpenDialog({
        title: 'Nevis - Import Archive',
        buttonLabel: 'Import',
        properties: ['openFile'],
        filters: [
          { name: 'Comma Separated Values', extensions: ['csv'] },
          { name: 'All Files', extensions: ['*'] },
        ],
      })
        .then(result => {
          if (!result.canceled) return this.$fs.readFile(result.filePaths[0], { encoding: 'utf8' })
          else throw Error()
        })
        .then(result => result.trim().split('\n'))
        .then(result => {
          if (!result[0].includes('CardNumber')) throw Error()
          else return result
        })
        .then(result => result.map(row => row.split(',')))
        .then(result => result.filter(row => !row[0].includes('CardNumber')))
        .then(result => result.map(row => [row[0], row[2].replace(/"/g, ''), row[3].replace(/"/g, ''), row[4].replace(/"/g, ''), this.getYearOfBirth(row[5]), row[6], row[7].replace(/"/g, '')]))
        .then(result => Promise.all([this.$archive.query('DELETE FROM people'), result]))
        .then(result => this.$archive.query('INSERT INTO people (siid, status, name, gender, yearOfBirth, membershipNumber, club) VALUES ?', [result[1]]))
        .then(result => this.$messages.addMessage(`${result.affectedRows} Archive Records Imported`))
        .catch(() => this.$messages.addMessage('Problem Importing the Archive', 'error'))
    },

    getYearOfBirth: function (date) {
      if (date.length === 4) return parseInt(date)
      else if (date.split('/')[2] && date.split('/')[2].length === 4) return parseInt(date.split('/')[2])
      else return 0
    },

    createTables: function () {
      const createDatabaseQueries = [
        `CREATE TABLE IF NOT EXISTS events(
          id INT AUTO_INCREMENT PRIMARY KEY,
          name TEXT,
          date TEXT)`,
        `CREATE TABLE IF NOT EXISTS courses(
          id INT AUTO_INCREMENT PRIMARY KEY,
          name TEXT,
          length INT,
          climb INT,
          type TEXT,
          controls TEXT,
          event INT,
          FOREIGN KEY (event) REFERENCES events(id) ON UPDATE CASCADE ON DELETE CASCADE)`,
        `CREATE TABLE IF NOT EXISTS competitors(
          id INT AUTO_INCREMENT PRIMARY KEY,
          name TEXT, siid TEXT,
          membershipNumber TEXT,
          ageClass TEXT,
          club TEXT,
          downloaded BOOLEAN,
          course INT,
          event INT,
          FOREIGN KEY (course) REFERENCES courses(id) ON UPDATE CASCADE ON DELETE SET NULL,
          FOREIGN KEY (event) REFERENCES events(id) ON UPDATE CASCADE ON DELETE CASCADE)`,
        `CREATE TABLE IF NOT EXISTS punches(
          id INT AUTO_INCREMENT PRIMARY KEY,
          time INT,
          controlCode TEXT,
          competitor INT,
          event INT,
          FOREIGN KEY (competitor) REFERENCES competitors(id) ON UPDATE CASCADE ON DELETE CASCADE,
          FOREIGN KEY (event) REFERENCES events(id) ON UPDATE CASCADE ON DELETE CASCADE)`,
      ]
      for (const query of createDatabaseQueries) {
        this.$database.query(query)
          .catch(() => this.$messages.addMessage('Problem Setting Up Database', 'error'))
      }
    },
  },
}
</script>
