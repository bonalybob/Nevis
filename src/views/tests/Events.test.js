import { mount, shallowMount } from '@vue/test-utils'
import Events from '@/views/Events'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockResolvedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - No Events', () => {
  const wrapper = shallowMount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockResolvedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Events', () => {
  const wrapper = shallowMount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockResolvedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  wrapper.setData({ events: [{ id: 1, name: 'Test', date: '1/2/3' }, { id: 2, name: 'Testing', date: '36/22/11' }, { id: 3, name: 'Event', date: '5/6/9' }] })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: false, query: jest.fn().mockResolvedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Events - Success', async () => {
  const wrapper = mount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockResolvedValue(['hello']) },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.events).toEqual(['hello'])
})

test('Get Events - Error', async () => {
  const wrapper = mount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockRejectedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.getEvents()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Events', 'error')
  expect(wrapper.vm.events.length).toBe(0)
})

test('Get Year Of Birth', () => {
  const wrapper = mount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockRejectedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  expect(wrapper.vm.getYearOfBirth('2001')).toBe(2001)
  expect(wrapper.vm.getYearOfBirth('1987')).toBe(1987)
  expect(wrapper.vm.getYearOfBirth('19/12/2014')).toBe(2014)
  expect(wrapper.vm.getYearOfBirth('1/2/2003')).toBe(2003)
  expect(wrapper.vm.getYearOfBirth('2001')).toBe(2001)
  expect(wrapper.vm.getYearOfBirth('5+/2-')).toBe(0)
})

test('Import Archive - Cancelled', async () => {
  const wrapper = mount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $fs: { readFile: jest.fn().mockResolvedValue() },
      $electron: { remote: { dialog: { showOpenDialog: jest.fn().mockResolvedValue({ canceled: true }) } } },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockRejectedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockRejectedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
  })
  await wrapper.vm.importArchive()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Importing the Archive', 'error')
})

test('Import Archive - Wrong File Type', async () => {
  const sampleArchiveForImport = `
CardNuber,CardLabel,CardStatus,Name,Sex,DateOfBirth,MemberNo,Club,Country
,,,Bob Eves,m,22/02/1995,TARP1001,St Lukes,GBR
,,,Jo Smith,f,12/12/1995,TARP1002,St Janes,IRL
`

  const wrapper = mount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $fs: { readFile: jest.fn().mockResolvedValue(sampleArchiveForImport) },
      $electron: {
        remote: { dialog: { showOpenDialog: jest.fn().mockResolvedValue({ canceled: false, filePaths: ['/test'] }) } },
      },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockRejectedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockResolvedValue(), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
    methods: {
      getYearOfBirth: jest.fn(() => 2000),
      createArchiveConnection: jest.fn(),
    },
  })
  await wrapper.vm.importArchive()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Importing the Archive', 'error')
})

test('Import Archive - Success', async () => {
  const sampleArchiveForImport = `
CardNumber,CardLabel,CardStatus,Name,Sex,DateOfBirth,MemberNo,Club,Country
,,,Bob Eves,m,22/02/1995,TARP1001,St Lukes,GBR
,,,Jo Smith,f,12/12/1995,TARP1002,St Janes,IRL
`

  const wrapper = mount(Events, {
    stubs: ['router-link'],
    mocks: {
      $mysql: { createConnection: jest.fn() },
      $fs: { readFile: jest.fn().mockResolvedValue(sampleArchiveForImport) },
      $electron: {
        remote: { dialog: { showOpenDialog: jest.fn().mockResolvedValue({ canceled: false, filePaths: ['/test'] }) } },
      },
      $database: { connection: { config: { host: '', port: '', user: '', password: '' } }, connected: true, query: jest.fn().mockResolvedValue() },
      $archive: { connection: {}, connected: true, query: jest.fn().mockResolvedValue({ affectedRows: 2 }), connect: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn(), clearMessages: jest.fn() },
    },
    methods: {
      getYearOfBirth: jest.fn(() => 2000),
      createArchiveConnection: jest.fn(),
    },
  })
  await wrapper.vm.importArchive()
  expect(wrapper.vm.$fs.readFile).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$fs.readFile).toHaveBeenLastCalledWith('/test', { encoding: 'utf8' })

  expect(wrapper.vm.$archive.query).toHaveBeenLastCalledWith('INSERT INTO people (siid, status, name, gender, yearOfBirth, membershipNumber,club) VALUES ?', [[
    ['', '', 'Bob Eves', 'm', 2000, 'TARP1001', 'St Lukes'],
    ['', '', 'Jo Smith', 'f', 2000, 'TARP1002', 'St Janes'],
  ]])
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('2 Archive Records Imported')
})
