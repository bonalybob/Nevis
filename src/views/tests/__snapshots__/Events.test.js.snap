// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Events 1`] = `
<main>
  <back-arrow-stub
    to="/"
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    Events
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/create"
    >
      Create Event
    </router-link-stub>
     
    <button
      class="button"
    >
      
      Import Archive
    
    </button>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <div
    class="my-shadow mx-12 mb-3 p-2"
  >
    <table
      class="w-full font-body"
    >
      <tr
        class="font-heading text-center hover:bg-blue-light"
      >
        <th>
          Event Name
        </th>
         
        <th>
          Event Date
        </th>
      </tr>
       
      <router-link-stub
        class="text-center even:bg-blue-lightest hover:bg-blue-light"
        tag="tr"
        to="/events/1"
      >
        <td>
          Test
        </td>
         
        <td>
          1/2/3
        </td>
      </router-link-stub>
      <router-link-stub
        class="text-center even:bg-blue-lightest hover:bg-blue-light"
        tag="tr"
        to="/events/2"
      >
        <td>
          Testing
        </td>
         
        <td>
          36/22/11
        </td>
      </router-link-stub>
      <router-link-stub
        class="text-center even:bg-blue-lightest hover:bg-blue-light"
        tag="tr"
        to="/events/3"
      >
        <td>
          Event
        </td>
         
        <td>
          5/6/9
        </td>
      </router-link-stub>
    </table>
  </div>
</main>
`;

exports[`Renders Correctly - No Events 1`] = `
<main>
  <back-arrow-stub
    to="/"
  />
   
  <h1
    class="mx-10 mb-1 px-1"
  >
    Events
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/create"
    >
      Create Event
    </router-link-stub>
     
    <button
      class="button"
    >
      
      Import Archive
    
    </button>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <!---->
</main>
`;
